File: LeetCodeExample.Test/Sorting/_00088_MergeSortedArray.py
# You are given two integer arrays nums1 and nums2, sorted in non-decreasing order, and two integers m and n, representing the number of elements in nums1 and nums2 respectively.

# Merge nums1 and nums2 into a single array sorted in non-decreasing order.

# The final sorted array should not be returned by the function, but instead be stored inside the array nums1. To accommodate this, nums1 has a length of m + n, where the first m elements denote the elements that should be merged, and the last n elements are set to 0 and should be ignored. nums2 has a length of n.

class Solution:
    def merge(self, nums1: List[int], m: int, nums2: List[int], n: int) -> None:
        """
        Do not return anything, modify nums1 in-place instead.
        """
        # Do it backwards
        i = len(nums1) - 1
        mi = m - 1
        ni = len(nums2) - 1
        while mi >= 0 and ni >= 0:
            if nums1[mi] >= nums2[ni]:
                # pick nums1
                nums1[i] = nums1[mi]
                i -= 1
                mi -= 1
            else:
                # pick nums2
                nums1[i] = nums2[ni]
                i -= 1
                ni -= 1
        
        # Do any remaining
        while mi >= 0:
            nums1[i] = nums1[mi]
            i -= 1
            mi -= 1

        while ni >= 0:
            nums1[i] = nums2[ni]
            i -= 1
            ni -= 1

File: LeetCodeExample.Test/Sorting/_00088_MergeSortedArray.cs
using NUnit.Framework;
using System;
using System.Collections.Generic;
using System.Linq;

namespace LeetCodeExample.Test
{
    /// <summary>
    //Given an array of integers nums and an integer target, return indices of the two numbers such that they add up to target.
    //You may assume that each input would have exactly one solution, and you may not use the same element twice.
    //You can return the answer in any order.
    /// </summary>
    public class _00088_MergeSortedArray
    {
        public void Merge(int[] nums1, int m, int[] nums2, int n)
        {
            int mi = m - 1;
            int ni = nums2.Length - 1;
            int i = nums1.Length - 1;
            // Interesting case. Let's do a merge sort but backwards
            // So we don't have to do any extra work
            while (mi >= 0 && ni >= 0)
            {
                if (nums1[mi] >= nums2[ni])
                {
                    // Pluck from nums1
                    nums1[i] = nums1[mi];
                    i--;
                    mi--;
                }
                else
                {
                    // Pluck from nums2
                    nums1[i] = nums2[ni];
                    ni--;
                    i--;
                }
            }
            // Do any remaining values
            while (mi >= 0)
            {
                // Pluck from nums1
                nums1[i] = nums1[mi];
                i--;
                mi--;
            }
            while (ni >= 0)
            {
                // Pluck from nums2
                nums1[i] = nums2[ni];
                i--;
                ni--;
            }
        }

        public void Merge2(int[] nums1, int m, int[] nums2, int n)
        {
            int i = 0;

            // Let's try not to create a new array
            // Move the elements to the end of the array nums1
            for (int j = nums1.Length - m; j < nums1.Length; j++)
            {
                nums1[j] = nums1[i];
                i++;
            }
            int mi = nums1.Length - m;
            int ni = 0;
            i = 0;
            //Console.WriteLine($"mi: {mi}, ni: {ni}");
            // Do a merge sort
            while (mi < nums1.Length && ni < n)
            {
                if (nums1[mi] <= nums2[ni])
                {
                    // Grab item from nums1
                    nums1[i] = nums1[mi];
                    i++;
                    mi++;
                }
                else
                {
                    // Grab item from nums2
                    nums1[i] = nums2[ni];
                    i++;
                    ni++;
                }
            }
            // Put in any remaining items
            while (mi < nums1.Length)
            {
                // Grab item from nums1
                nums1[i] = nums1[mi];
                i++;
                mi++;
            }

            while (ni < n && i < nums1.Length)
            {
                // Grab item from nums1
                nums1[i] = nums2[ni];
                i++;
                ni++;
            }
        }
    }
}